// forma-embedded-view-sdk/dist/internal/scene/terrain.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _GroundTextureApi_iframeMessenger;
export class TerrainApi {
    constructor(iframeMessenger) {
        this.groundTexture = new GroundTextureApi(iframeMessenger);
    }
}
class GroundTextureApi {
    constructor(iframeMessenger) {
        _GroundTextureApi_iframeMessenger.set(this, void 0);
        __classPrivateFieldSet(this, _GroundTextureApi_iframeMessenger, iframeMessenger, "f");
    }
    async add({ name, canvas, position, scale }) {
        const canvasUrl = canvas.toDataURL();
        await __classPrivateFieldGet(this, _GroundTextureApi_iframeMessenger, "f").sendRequest("scene/terrain/ground-texture/add", {
            name,
            canvasUrl,
            position,
            scale,
        });
    }
    async updateTextureData({ name, canvas, }) {
        const canvasUrl = canvas.toDataURL();
        await __classPrivateFieldGet(this, _GroundTextureApi_iframeMessenger, "f").sendRequest("scene/terrain/ground-texture/update-texture-data", {
            name,
            canvasUrl,
        });
    }
    async updatePosition({ name, position, }) {
        await __classPrivateFieldGet(this, _GroundTextureApi_iframeMessenger, "f").sendRequest("scene/terrain/ground-texture/update-position", { name, position });
    }
    async remove({ name }) {
        await __classPrivateFieldGet(this, _GroundTextureApi_iframeMessenger, "f").sendRequest("scene/terrain/ground-texture/remove", {
            name,
        });
    }
}
_GroundTextureApi_iframeMessenger = new WeakMap();
